util/linuxfw/iptables_runner.go

```diff
@@ -0,0 +1,485 @@
+// is a wrapper to *iptables.IPTables.
+func NewIptablesRunner(logf logger.Logf) (*iptablesRunner, error) {
+	ipt4, err := iptables.NewWithProtocol(iptables.ProtocolIPv4)
+	if err != nil {
+		return nil, err
+	}
+
+	supportsV6, supportsV6NAT := false, false
+	v6err := checkIPv6(logf)
+	if v6err != nil {
+		logf("disabling tunneled IPv6 due to system IPv6 config: %v", v6err)
+	} else {
+		supportsV6 = true
+		supportsV6NAT = supportsV6 && checkSupportsV6NAT()
+		logf("v6nat = %v", supportsV6NAT)
+	}
+
+	var ipt6 *iptables.IPTables
+	if supportsV6 {
+		ipt6, err = iptables.NewWithProtocol(iptables.ProtocolIPv6)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return newIptablesRunnerAdvanced(ipt4, ipt6, supportsV6, supportsV6NAT), nil
+}
+
+// newIptablesRunnerAdvanced returns an iptablesRunner.
+// It is a wrapper for iptablesRunner init, which exists for testing purpose."
+func newIptablesRunnerAdvanced(ipt4, ipt6 iptablesInterface, supV6, supV6NAT bool) *iptablesRunner {
+	return &iptablesRunner{ipt4, ipt6, supV6, supV6NAT}
+}
+
+// GetV6Available returns true if the system supports IPv6.
+func (i *iptablesRunner) GetV6Available() bool {
+	return i.v6Available
+}
+
+// GetV6NATAvailable returns true if the system supports IPv6 NAT.
+func (i *iptablesRunner) GetV6NATAvailable() bool {
+	return i.v6NATAvailable
+}
+
+func isErrChainNotExist(err error) bool {
+	return errCode(err) == 1
+}
+
+// AddLoopbackRule adds a iptables rule to permit loopback traffic to
+// a local Tailscale IP.
+func (i *iptablesRunner) AddLoopbackRule(addr netip.Addr) error {
+	nf := i.ipt4
+	if addr.Is6() {
+		nf = i.ipt6
+	}
+
+	if err := nf.Insert("filter", "ts-input", 1, "-i", "lo", "-s", addr.String(), "-j", "ACCEPT"); err != nil {
+		return fmt.Errorf("adding loopback allow rule for %q: %w", addr, err)
+	}
+
+	return nil
+}
+
+// tsChain returns the name of the tailscale sub-chain corresponding
+// to the given "parent" chain (e.g. INPUT, FORWARD, ...).
+func tsChain(chain string) string {
+	return "ts-" + strings.ToLower(chain)
+}
+
+// getIptablebyProto returns an object you can use to manipulate IP tables for the given protocol.
+func (i *iptablesRunner) getIptablebyProto(proto ProtoVersion) iptablesInterface {
+	switch proto {
+	case IPv4:
+		return i.ipt4
+	case IPv6:
+		return i.ipt6
+	default:
+		panic(fmt.Sprintf("unknown protocol %v", proto))
+	}
+}
+
+// DelLoopbackRule removes the iptables rule permitting loopback
+// traffic to a Tailscale IP.
+func (i *iptablesRunner) DelLoopbackRule(addr netip.Addr) error {
+	nf := i.ipt4
+	if addr.Is6() {
+		nf = i.ipt6
+	}
+
+	if err := nf.Delete("filter", "ts-input", "-i", "lo", "-s", addr.String(), "-j", "ACCEPT"); err != nil {
+		return fmt.Errorf("deleting loopback allow rule for %q: %w", addr, err)
+	}
+
+	return nil
+}
+
+// getTables gets the available iptablesInterface in iptables runner.
+func (i *iptablesRunner) getTables() []iptablesInterface {
```

Comment:
If we had a sibling getNATTables() then we could apply this pattern to a lot of cases below that are conditional on NAT, too.

Reply:
Sounds good